lib/galaxy/web/security/__init__.py

```diff
@@ -127,4 +133,15 @@ def __init__(self, secret_base):
         self.secret_base = secret_base
 
     def __missing__(self, key):
-        return Blowfish.new(self.secret_base + "__" + key)
+        assert len(key) < 15, KIND_TOO_LONG_MESSAGE
+        secret = self.secret_base + "__" + key
+        return Blowfish.new(_last_bits(secret))
+
+
+def _last_bits(secret):
+    """We append the kind at the end, so just use the bits at the end.
+    """
+    last_bits = secret
+    if len(secret) > MAXIMUM_ID_SECRET_LENGTH:
+        last_bits = secret[-MAXIMUM_ID_SECRET_LENGTH:]
```

Comment:
This won't work on Python3, were the length of a string containing UTF-8 characters is different from the number of bytes, e.g.:
```
$ python2
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> secret = "◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎"
>>> print len(secret)
57
>>> 
$ python3
Python 3.4.3 (default, Nov 17 2016, 01:08:31) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> secret = "◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎"
>>> print(len(secret))
19
>>> 
```

I'd rewrite these lines as:
```python
    last_bits = smart_str(secret)
    if len(last_bits) > MAXIMUM_ID_SECRET_LENGTH:
        last_bits = last_bits[-MAXIMUM_ID_SECRET_LENGTH:]
```

and add
```python
from galaxy.util import smart_str
```
in the import section.

Reply:
@nsoranzo Thanks - what do you think about 7f84f8e instead so it is more clear we are only modifying ``id_secret`` if it is too many bits?

Reply:
https://github.com/galaxyproject/galaxy/commit/7f84f8e801654bf04b8b920de5de18a287f9ca14 would work only if we can guarantee that `secret` is always of type `str` in Python2:
```
$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from Crypto.Cipher import Blowfish
>>> from galaxy.util import smart_str
>>> MAXIMUM_ID_SECRET_LENGTH = 56
>>> secret = u"◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎"
>>> print type(secret)
<type 'unicode'>
>>> print len(secret)
18
>>> last_bits = secret
>>> as_str = smart_str(secret)
>>> len(as_str)
54
>>> len(as_str) > MAXIMUM_ID_SECRET_LENGTH
False
>>> Blowfish.new(last_bits)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/.venv/local/lib/python2.7/site-packages/Crypto/Cipher/Blowfish.py", line 101, in new
    return BlowfishCipher(key, *args, **kwargs)
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/.venv/local/lib/python2.7/site-packages/Crypto/Cipher/Blowfish.py", line 65, in __init__
    blockalgo.BlockAlgo.__init__(self, _Blowfish, key, *args, **kwargs)
  File "/usr/users/ga002/soranzon/software/nsoranzo_galaxy/.venv/local/lib/python2.7/site-packages/Crypto/Cipher/blockalgo.py", line 141, in __init__
    self._cipher = factory.new(key, *args, **kwargs)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-17: ordinal not in range(128)
```

Reply:
Okay - I'll go with your variant as long you promise smart_str isn't going to break anyone's existing values. It looks like existing unicode secrets wouldn't work without this tweak anyway so that probably isn't an issue?

Reply:
```
$ python
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from Crypto.Cipher import Blowfish
>>> from galaxy.util import smart_str
>>> secret = "◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎◎"
>>> b_old = Blowfish.new(secret)
>>> s = 'foobar12'
>>> old_enc_s = b_old.encrypt(s)
>>> b_new = Blowfish.new(smart_str(secret))
>>> print b_new.encrypt(s) == old_enc_s
True
```

Reply:
Fair enough - sorry for being dense. I've rebased it all into one commit including this change.